File: icc-x-api/utils/lzma-wasm.ts
/**
 * WASM-based LZMA2/XZ compression using liblzma compiled via Emscripten.
 * The WASM module is lazy-loaded on first use.
 */

interface LzmaWasmModule {
  _lzma_wasm_compress(ptr: number, len: number, preset: number): number
  _lzma_wasm_decompress(ptr: number, len: number): number
  _get_result_ptr(): number
  _get_result_len(): number
  _free_result(): void
  _malloc(size: number): number
  _free(ptr: number): void
  HEAPU8: Uint8Array
}

type ModuleFactory = () => Promise<LzmaWasmModule>

let modulePromise: Promise<LzmaWasmModule> | null = null

function getModule(): Promise<LzmaWasmModule> {
  if (!modulePromise) {
    modulePromise = (async () => {
      // Dynamic import of the Emscripten glue code
      const factory: ModuleFactory = (await import('../wasm/lzma/lzma.js')).default
      return factory()
    })()
  }
  return modulePromise
}

export async function lzmaWasmCompress(data: ArrayBuffer | Uint8Array, preset: number = 6): Promise<ArrayBuffer> {
  const mod = await getModule()
  const input = data instanceof ArrayBuffer ? new Uint8Array(data) : data

  const inputPtr = mod._malloc(input.length)
  if (!inputPtr) throw new Error('WASM malloc failed for input')

  try {
    mod.HEAPU8.set(input, inputPtr)
    const ret = mod._lzma_wasm_compress(inputPtr, input.length, preset)
    if (ret !== 0) {
      mod._free_result()
      throw new Error(`LZMA WASM compress failed with code ${ret}`)
    }

    const resultPtr = mod._get_result_ptr()
    const resultLen = mod._get_result_len()
    const output = new Uint8Array(resultLen)
    output.set(mod.HEAPU8.subarray(resultPtr, resultPtr + resultLen))
    mod._free_result()
    return output.buffer
  } finally {
    mod._free(inputPtr)
  }
}

export async function lzmaWasmDecompress(data: ArrayBuffer | Uint8Array): Promise<ArrayBuffer> {
  const mod = await getModule()
  const input = data instanceof ArrayBuffer ? new Uint8Array(data) : data

  const inputPtr = mod._malloc(input.length)
  if (!inputPtr) throw new Error('WASM malloc failed for input')

  try {
    mod.HEAPU8.set(input, inputPtr)
    const ret = mod._lzma_wasm_decompress(inputPtr, input.length)
    if (ret !== 0) {
      mod._free_result()
      throw new Error(`LZMA WASM decompress failed with code ${ret}`)
    }

    const resultPtr = mod._get_result_ptr()
    const resultLen = mod._get_result_len()
    const output = new Uint8Array(resultLen)
    output.set(mod.HEAPU8.subarray(resultPtr, resultPtr + resultLen))
    mod._free_result()
    return output.buffer
  } finally {
    mod._free(inputPtr)
  }
}

File: icc-x-api/wasm/lzma/lzma_wasm.c
#include <lzma.h>
#include <stdlib.h>
#include <string.h>
#include <emscripten/emscripten.h>

/* Result buffer shared between calls — caller must copy before next call. */
static uint8_t *result_buf = NULL;
static size_t result_len = 0;

EMSCRIPTEN_KEEPALIVE
size_t get_result_len(void) {
    return result_len;
}

EMSCRIPTEN_KEEPALIVE
uint8_t *get_result_ptr(void) {
    return result_buf;
}

EMSCRIPTEN_KEEPALIVE
void free_result(void) {
    free(result_buf);
    result_buf = NULL;
    result_len = 0;
}

/*
 * Compress `input` (length `input_len`) using LZMA2 in the .xz container format.
 * `preset` maps to liblzma presets (0-9, 6 is default).
 * Returns 0 on success, non-zero on error.
 * After success, call get_result_ptr() / get_result_len() to read the output.
 */
EMSCRIPTEN_KEEPALIVE
int lzma_wasm_compress(const uint8_t *input, size_t input_len, uint32_t preset) {
    free_result();

    /* Worst-case: compressed can be larger than input. Allocate generously. */
    size_t out_capacity = input_len + input_len / 3 + 256;
    result_buf = (uint8_t *)malloc(out_capacity);
    if (!result_buf) return 1;

    size_t out_pos = 0;
    lzma_ret ret = lzma_easy_buffer_encode(
        preset, LZMA_CHECK_CRC64,
        NULL, /* allocator */
        input, input_len,
        result_buf, &out_pos, out_capacity
    );

    if (ret != LZMA_OK) {
        free_result();
        return (int)ret;
    }

    result_len = out_pos;
    return 0;
}

/*
 * Decompress .xz data.
 * Returns 0 on success, non-zero on error.
 */
EMSCRIPTEN_KEEPALIVE
int lzma_wasm_decompress(const uint8_t *input, size_t input_len) {
    free_result();

    /* Start with 4x input as estimate, grow if needed. */
    size_t out_capacity = input_len * 4;
    if (out_capacity < 1024) out_capacity = 1024;

    lzma_stream strm = LZMA_STREAM_INIT;
    lzma_ret ret = lzma_stream_decoder(&strm, UINT64_MAX, 0);
    if (ret != LZMA_OK) return (int)ret;

    result_buf = (uint8_t *)malloc(out_capacity);
    if (!result_buf) {
        lzma_end(&strm);
        return 1;
    }

    strm.next_in = input;
    strm.avail_in = input_len;
    strm.next_out = result_buf;
    strm.avail_out = out_capacity;

    while (1) {
        ret = lzma_code(&strm, LZMA_FINISH);
        if (ret == LZMA_STREAM_END) {
            break;
        }
        if (ret == LZMA_OK && strm.avail_out == 0) {
            /* Need more output space — double the buffer. */
            size_t used = out_capacity - strm.avail_out;
            out_capacity *= 2;
            uint8_t *tmp = (uint8_t *)realloc(result_buf, out_capacity);
            if (!tmp) {
                lzma_end(&strm);
                free_result();
                return 1;
            }
            result_buf = tmp;
            strm.next_out = result_buf + used;
            strm.avail_out = out_capacity - used;
            continue;
        }
        if (ret != LZMA_OK) {
            lzma_end(&strm);
            free_result();
            return (int)ret;
        }
    }

    result_len = strm.total_out;
    lzma_end(&strm);
    return 0;
}